### ISSUE TITLE
All tests are failing for 3.11.4.0, with `Could not initialize class com.datastax.driver.core.CCMBridge (java.lang.NoClassDefFoundError)` error

### ISSUE THREAD

**Issue by dimakr** [scylladb/scylla-java-driver-matrix#56]
All tests are failing with `Could not initialize class com.datastax.driver.core.CCMBridge (java.lang.NoClassDefFoundError)` error.

The issue was noticed during 2024.1.12 verification cycle, but as per scylla-master builds of the test it started from 2024-10-10 - first time observed in [java-driver-matrix-test build #1455](https://argus.scylladb.com/tests/driver-matrix-tests/f323f9cc-67f8-4d51-8f8f-7327ff3c0883).

Probably caused by the patch https://github.com/scylladb/scylla-java-driver-matrix/commit/0572bbef5d266b5be65a2f6e7acd7fc0206c332b, the last patch in the repo which was merged on 2024-10-10.

**Comment by Bouncheck** [comment]
Looks like CCMBridge cannot initialize because it cannot parse scylla version in the form of a hash. It expects either an enterprise or OSS version number. I'm not sure why this did not surface earlier since the patch from 10/10 just changes the code of a single test.

**Comment by Bouncheck** [comment]
https://github.com/scylladb/scylla-java-driver-matrix/blob/0572bbef5d266b5be65a2f6e7acd7fc0206c332b/scripts/jenkins-pipelines/groovy_modules/test.groovy#L72

@dimakr do you know if this script should have `args.ScyllaVersion` instead of `{env.SCYLLA_VERSION}` here? It looks like correct ScyllaVersion is usually passed, but later on the hash is used anyway. For reference I've been looking at https://jenkins.scylladb.com/view/master/job/scylla-master/job/driver-tests/job/java-driver-matrix-test/1504/consoleFull

**Comment by fruch** [comment]
@Bouncheck 

https://argus.scylladb.com/tests/driver-matrix-tests/f323f9cc-67f8-4d51-8f8f-7327ff3c0883
right now, all of the release have broken patches

maybe it should be attend first

**Comment by Bouncheck** [comment]
My assumption was that for example tests foe 3.11.5.3 will apply only the patch file from 3.11.5.3 folder if it exists. Seems like it applies earlier patches as well.

**Comment by fruch** [comment]
> My assumption was that for example tests for 3.11.5.3 will apply only the patch file from 3.11.5.3 folder if it exists. Seems like it applies earlier patches as well.

you first assumption is correct, it would apply only one patch file
the assumption the older ones aren't needed anymore, was incorrect.

the fact I approved with asking for prove of it running, was also incorrect...

**Comment by Bouncheck** [comment]
Wait, I'm a little confused

> > My assumption was that for example tests for 3.11.5.3 will apply only the patch file from 3.11.5.3 folder if it exists. Seems like it applies earlier patches as well.
> 
> you first assumption is correct, it would apply only one patch file the assumption the older ones aren't needed anymore, was incorrect.

But when I look at the log of run 1504 I see:
```
00:09:36  [notice] To update, run: pip install --upgrade pip
00:09:36  INFO:root:=== Going to test those versions: ['3.11.5.3', '3.11.4.0', '4.18.0.1', '4.17.0.1']
00:09:36  INFO:root:=== JAVA DRIVER VERSION 3.11.5.3 ===
00:09:36  INFO:root:Execute the cmd 'git checkout .'
00:09:36  INFO:root:Execute the cmd 'git checkout 3.11.5.3'
00:09:36  INFO:root:The full directory for '3.11.5.3' tag is '/jenkins/workspace/scylla-master/driver-tests/java-driver-matrix-test/scylla-java-driver-matrix/versions/scylla/3.11.5.3'
00:09:36  INFO:root:Show patch's statistics for file '/jenkins/workspace/scylla-master/driver-tests/java-driver-matrix-test/scylla-java-driver-matrix/versions/scylla/3.11.4.0/patch'
00:09:36  INFO:root:Execute the cmd 'git apply --stat /jenkins/workspace/scylla-master/driver-tests/java-driver-matrix-test/scylla-java-driver-matrix/versions/scylla/3.11.4.0/patch'
00:09:36   .../core/LoadBalancingPolicyBootstrapTest.java     |   10 +++++++++-
00:09:36   1 file changed, 9 insertions(+), 1 deletion(-)
00:09:36  INFO:root:Detect patch's errors for file '/jenkins/workspace/scylla-master/driver-tests/java-driver-matrix-test/scylla-java-driver-matrix/versions/scylla/3.11.4.0/patch'
00:09:36  INFO:root:Execute the cmd 'git apply -v --check /jenkins/workspace/scylla-master/driver-tests/java-driver-matrix-test/scylla-java-driver-matrix/versions/scylla/3.11.4.0/patch'
00:09:36  ERROR:root:Failed to apply patch '/jenkins/workspace/scylla-master/driver-tests/java-driver-matrix-test/scylla-java-driver-matrix/versions/scylla/3.11.4.0/patch' to version '3.11.5.3', with: 'Checking patch driver-core/src/test/java/com/datastax/driver/core/LoadBalancingPolicyBootstrapTest.java...
00:09:36  error: while searching for:
00:09:36        ccm().stop(nodeToStop);
00:09:36        ccm().waitForDown(nodeToStop);
00:09:36  
00:09:36        HistoryPolicy policy = new HistoryPolicy(new RoundRobinPolicy());
00:09:36        Cluster cluster = register(createClusterBuilder().withLoadBalancingPolicy(policy).build());
00:09:36  
00:09:36        try {
00:09:36          cluster.init();
00:09:36  
00:09:36  error: patch failed: driver-core/src/test/java/com/datastax/driver/core/LoadBalancingPolicyBootstrapTest.java:104
00:09:36  error: driver-core/src/test/java/com/datastax/driver/core/LoadBalancingPolicyBootstrapTest.java: patch does not apply
00:09:36  '
00:09:36  ERROR:root:3.11.5.3 failed
```
Is this not the case of the test run for 3.11.5.3 trying to apply the patch for 3.11.4.0 I've added? Is this supposed or not supposed to happen?

**Comment by fruch** [comment]
> Wait, I'm a little confused
> 
> > > My assumption was that for example tests for 3.11.5.3 will apply only the patch file from 3.11.5.3 folder if it exists. Seems like it applies earlier patches as well.
> > 
> > you first assumption is correct, it would apply only one patch file the assumption the older ones aren't needed anymore, was incorrect.
> 
> But when I look at the log of run 1504 I see:
> ```
> 00:09:36  [notice] To update, run: pip install --upgrade pip
> 00:09:36  INFO:root:=== Going to test those versions: ['3.11.5.3', '3.11.4.0', '4.18.0.1', '4.17.0.1']
> 00:09:36  INFO:root:=== JAVA DRIVER VERSION 3.11.5.3 ===
> 00:09:36  INFO:root:Execute the cmd 'git checkout .'
> 00:09:36  INFO:root:Execute the cmd 'git checkout 3.11.5.3'
> 00:09:36  INFO:root:The full directory for '3.11.5.3' tag is '/jenkins/workspace/scylla-master/driver-tests/java-driver-matrix-test/scylla-java-driver-matrix/versions/scylla/3.11.5.3'
> 00:09:36  INFO:root:Show patch's statistics for file '/jenkins/workspace/scylla-master/driver-tests/java-driver-matrix-test/scylla-java-driver-matrix/versions/scylla/3.11.4.0/patch'
> 00:09:36  INFO:root:Execute the cmd 'git apply --stat /jenkins/workspace/scylla-master/driver-tests/java-driver-matrix-test/scylla-java-driver-matrix/versions/scylla/3.11.4.0/patch'
> 00:09:36   .../core/LoadBalancingPolicyBootstrapTest.java     |   10 +++++++++-
> 00:09:36   1 file changed, 9 insertions(+), 1 deletion(-)
> 00:09:36  INFO:root:Detect patch's errors for file '/jenkins/workspace/scylla-master/driver-tests/java-driver-matrix-test/scylla-java-driver-matrix/versions/scylla/3.11.4.0/patch'
> 00:09:36  INFO:root:Execute the cmd 'git apply -v --check /jenkins/workspace/scylla-master/driver-tests/java-driver-matrix-test/scylla-java-driver-matrix/versions/scylla/3.11.4.0/patch'
> 00:09:36  ERROR:root:Failed to apply patch '/jenkins/workspace/scylla-master/driver-tests/java-driver-matrix-test/scylla-java-driver-matrix/versions/scylla/3.11.4.0/patch' to version '3.11.5.3', with: 'Checking patch driver-core/src/test/java/com/datastax/driver/core/LoadBalancingPolicyBootstrapTest.java...
> 00:09:36  error: while searching for:
> 00:09:36        ccm().stop(nodeToStop);
> 00:09:36        ccm().waitForDown(nodeToStop);
> 00:09:36  
> 00:09:36        HistoryPolicy policy = new HistoryPolicy(new RoundRobinPolicy());
> 00:09:36        Cluster cluster = register(createClusterBuilder().withLoadBalancingPolicy(policy).build());
> 00:09:36  
> 00:09:36        try {
> 00:09:36          cluster.init();
> 00:09:36  
> 00:09:36  error: patch failed: driver-core/src/test/java/com/datastax/driver/core/LoadBalancingPolicyBootstrapTest.java:104
> 00:09:36  error: driver-core/src/test/java/com/datastax/driver/core/LoadBalancingPolicyBootstrapTest.java: patch does not apply
> 00:09:36  '
> 00:09:36  ERROR:root:3.11.5.3 failed
> ```
> Is this not the case of the test run for 3.11.5.3 trying to apply the patch for 3.11.4.0 I've added? Is this supposed or not supposed to happen?
> 

Since there's no patch for 3.11.5.4 it would fall back to the last available, in some cases it would just work, and we can keep at it until patch fails, 

And then we need to rebuild the patch after fixing the conflict

**Comment by Bouncheck** [comment]
I see, so my mistake was assuming there was no fallback. So the patch for 3.11.4.0 should've been made on top of the one for 3.7.1.0.
To sum up I should remake that patch on top of the patch for 3.7.1.0 (newest before version 3.11.4.0) and then separately make a patch (fix conflicts) for 3.11.5.3 if it's still broken.

**Comment by fruch** [comment]
in most case it would be to drag the old patch into the current borken patch, and fix it until it working

go to older patches and changing them doesn't make lots of sense, since we are running last two version.
and change the old version might break them (and we aren't testing)

old version are mainly for reference, if we need to run and compere some tests.

yet again, in case of your original change it might as well fit older releases, and the flow you suggest would work
